src/core/NewDomain.ml

```diff
@@ -0,0 +1,3041 @@
+
+      (* This is essentially C_M in [F]. *)
+      let cap_frame =
+        Cap
+          {r = fhcom.r;
+           r' = fhcom.r';
+           ty = fhcom.cap;
+           sys = fhcom.sys}
+      in
+
+      (* This serves as `O` and the diagonal face in [F]
+       * for the coherence conditions in `fhcom.sys` and `s=s'`. *)
+      let hcom_template rel y_dest ty =
+        make_hcom rel r y_dest ~ty
+          ~cap:(Val.run rel cap)
+          ~sys:(ConAbsSys.run rel sys)
+      in
+
+      (* This is `P` in [F]. *)
+      let cap = Val.make @@
+        let ty = Val.unleash fhcom.cap in
+        let cap = Val.plug rel cap_frame cap in
+        let sys =
+          let ri_faces = ConAbsSys.plug rel cap_frame sys in
+          let si_faces = ConAbsSys.foreach_gen fhcom.sys @@ fun si s'i abs ->
+            let rel = Rel.equate' si s'i rel in
+            let cap_frame = Frame.run rel cap_frame in
+            ConAbs.bind @@ fun y ->
+            (* XXX FIXME This is not the most efficient code because we know what [cap_frame]
+             * will reduce to, but maybe we can afford this? *)
+            Con.plug rel cap_frame @@ hcom_template rel y (ConAbs.inst rel abs s')
+          in
+          let diag =
+            r, r',
+            LazyValAbs.bind @@ fun y ->
+            let rel = Rel.equate' r r' rel in
+            hcom_template rel y (Val.run_then_unleash rel fhcom.cap)
+          in
+          ConAbsSys.force rel [diag] @ ri_faces @ si_faces
+        in
+        rigid_hcom rel r r' ~ty ~cap ~sys
+      in
+      let sys =
+        ConSys.foreach_gen fhcom.sys @@ fun si s'i abs ->
+        let rel = Rel.equate' si s'i rel in
+        hcom_template rel r' (ConAbs.inst rel abs s')
+      in
+      Box {r; r'; cap; sys}
+
+    | Neu info ->
+      let neu =
+        DelayedNeu.make
+          {head = NHCom {r; r'; ty = info.neu; cap; sys};
+           frames = Emp}
+      in
+      let neu_sys =
+        let cap_face = r, r', LazyVal.make_from_lazy @@ lazy begin Val.unleash cap end in
+        let tube_faces =
+          ConSys.foreach_gen sys @@ fun s s' abs ->
+          let rel' = Rel.equate' s s' rel in
+          ConAbs.inst rel' abs r'
+        in
+        let old_faces =
+          ListUtil.foreach info.neu.sys @@ ConFace.gen @@ fun s s' bdy ->
+          let rel' = Rel.equate' s s' rel in
+          let ty = run rel' bdy in
+          let cap = Val.run rel' cap in
+          let sys = ConAbsSys.run rel' sys in
+          make_hcom rel' r r' ~ty ~cap ~sys
+        in
+        cap_face :: tube_faces @ old_faces
+      in
+      Neu {ty = Val.make ty; neu = {neu; sys = neu_sys}}
+
+    | _ ->
+      raise PleaseRaiseProperError
+
+
+  and rigid_multi_coe r r' data_abs args =
+    let Abs (x, ty) = data_abs in
+    (* Is there a way to write this code where I don't need to look up the argument specs in the signature? I hope so...
+       Otherwise, I need to thrad the global environment through too much stuff.
```

Comment:
I think what I should do is bung into `Data` a closure of the `constrs` together with an **global** environment which can be used to evaluate them. It won't be necessary for it to respond to `run` or substitution or anything like that, because it is global.